book/4-processing.md

```diff
@@ -13,3 +13,315 @@ kernelspec:
+
+    Raises
+    ------
+    ValueError
+        If the arrays don't have the same shape or chunks.
+    """
+    if array_1.shape != array_2.shape:
+        raise ValueError(
+            f"Input shape ({array_1.shape}) != output shape {array_2.shape}"
+        )
+    if array_1.chunks != array_2.chunks:
+        raise ValueError(
+            f"Input chunk ({array_1.chunks}) != output chunks {array_2.chunks}"
+        )
+```
+
+Next we're going to put all these functions together to make a function that sets up a number of jobs.
+Each job will read a single chunk of data from an input array, apply our function, and then write the result to a single chunk of the output array.
+
+This uses the concept of a 'delayed' function, taken from the [`joblib`](https://joblib.readthedocs.io) Python library.
+When a delayed function is called, it is not run, but instead a reference to the function and the arguments to be applied to it are saved.
+Lets see how this works with a simple example:
+
+```{code-cell} ipython3
+import joblib
+
+def add(x: int, *, y: int) -> int:
+    print(f"Adding {x} to {y}")
+    return x + y
+
+delayed_add = joblib.delayed(add)
+job = delayed_add(3, y=4)
+print(job)
+```
+
+We can tell that because nothing has been printed, the function hasn't been run.
+The result is a tuple, containing the function, any arguments, and any keyword arguments to be applied to it.
+We can run the job manually:
+
+```{code-cell} ipython3
+function, args, kwargs = job
+function(*args, **kwargs)
+```
+
+Alternatively `joblib` has a builtin class to execute many jobs in parallel:
+
+```{code-cell} ipython3
+jobs = [
+    delayed_add(3, y=4),
+    delayed_add(2, y=1)
+]
+executor = joblib.Parallel(n_jobs=1)
+results = executor(jobs)
+print(results)
+```
+
+With knowledge of delayed function, we can create a delayed function to process individual slices of an array
+
+```{code-cell} ipython3
+delayed_apply_to_slice = joblib.delayed(apply_to_slice)
+```
+
+And then use this to create a function that will set up jobs to apply another function across every chunk of a Zarr array
+
+```{code-cell} ipython3
+def elementwise_jobs(
+    f: Callable[[npt.NDArray[Any]], npt.NDArray[Any]],
+    *,
+    input_array: zarr.Array,
+    output_array: zarr.Array,
+) -> list[joblib.delayed]:
+    """
+    Apply a function to all chunks of a Zarr array.
+
+    Parameters
+    ----------
+    f : 
+        Function to apply to each chunk.
+    input_array :
+        Array to read from.
+    output_array :
+        Array to write to.
+    """
+    check_same_shape(input_array, output_array)
+    return [
+        delayed_apply_to_slice(f, input_array, output_array, slc)
+        for slc in all_chunk_slices(output_array)
+    ]
+```
+
+Lets make some use of this function, to threshold an image using a thresholding function.
+
+```{code-cell} ipython3
+import numpy as np
+
+def threshold(array):
+    return np.clip(array, 128, 256)
```

Comment:
I would expect "thresholding" to generate a binary image, not clipping to a range of values?